lucene/core/src/java/org/apache/lucene/index/AutomatonTermsEnum.java

```diff
@@ -54,18 +56,20 @@
   private final boolean finite;
   // array of sorted transitions for each state, indexed by state number
   private final Automaton automaton;
-  // for path tracking: each long records gen when we last
+  // for path tracking: each short records gen when we last
   // visited the state; we use gens to avoid having to clear
-  private final long[] visited;
```

Comment:
Main change is replacing the long[] by a short[] to reduce memory usage (the size is the number of automaton states). I also tried a FixedBitSet but it impacts perf.

Reply:
visited-state-tracking is only needed when the automaton accepts an infinite language. We use it for loop detection. I think before we get too fancy with how we clear it, we should first stop being stupid about it?

So it is wasteful that we do this stuff when `finite == true` (example: fuzzy query) because we will never even look for a loop. its just that the current code unconditionally records states that it visited.

I think first, in the ctor when `finite == true`, `visited[]` can be initialized to `null` or `new long[0]` or something, and we change this line:
```
visited[state] = curGen;
```
to something like this:
```
if (!finite)
  visited[state] = curGen;
```

I agree we should separately avoid tracking 64 bits per state when only 1 is needed. But before optimizing the storage, first lets avoid doing this stuff at all for ones like complex fuzzy queries?

Reply:
Good catch.